### PR TITLE
A cop to check erroneous next within reduce.

### PR DESCRIPTION
**Pull request by mvidner**
In a reduce (aka inject), the accumulator value must be maintained.
If you want to short-circuit an iteration and do it with a `next`
the accumulator will become `nil`. You should use `next acc`.

For reference, here is a real bug reported in my team about such code:
http://lists.opensuse.org/yast-devel/2015-11/msg00094.html

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -170,6 +170,7 @@
 * `Style/Documentation` recognizes 'Constant = Class.new' as a class definition. ([@alexdowad][])
 * [#1608](https://github.com/bbatsov/rubocop/issues/1608): Add new 'align_braces' style for `Style/IndentHash`. ([@alexdowad][])
 * `Style/Next` can autocorrect. ([@alexdowad][])
+* [#????](https://github.com/bbatsov/rubocop/pull/????): New cop `Lint/NextWithoutAccumulator` finds bare `next` in `reduce`/`inject` blocks which assigns `nil` to the accumulator. ([@mvidner][])
```
Comment (alexdowad) [inline CHANGELOG.md]: ???? needs to be fixed.

File: lib/rubocop/cop/lint/next_without_accumulator.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+          _, method_name, method_args = *method
+          return unless METHODS.include? method_name
+          return if method_args && method_args.type == :sym
+
+          return false unless body.type == :begin
+
+          # Astrolabe::Node#each_node FTW!
+          # (actually we want a recursive detect
+          # but astrolabe cannot do a short circuit on a match)
+          found_void_next = nil
```
Comment (alexdowad) [inline lib/rubocop/cop/lint/next_without_accumulator.rb]: This is not necessary. Once you find a "void next" node, you can call `add_offense` and then just `return`.

File: lib/rubocop/cop/lint/next_without_accumulator.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      #   result = (1..4).reduce(0) do |acc, i|
+      #     next acc if i.odd?
+      #     acc + i
+      #   end
+      class NextWithoutAccumulator < Cop
+        MSG = 'Use `next` with an accumulator argument in a `reduce`.'.freeze
+        METHODS = [:inject, :reduce]
+
+        def on_block(node)
+          method, _args, body = *node
```
Comment (alexdowad) [inline lib/rubocop/cop/lint/next_without_accumulator.rb]: Lines 25-34 can be compressed into one of our "node patterns". The pattern you need should be: `(block (send _ {:reduce :inject} !sym) _ begin)`

File: lib/rubocop/cop/lint/next_without_accumulator.rb (hunk 2 above)
Comment (alexdowad) [inline lib/rubocop/cop/lint/next_without_accumulator.rb]: Or better yet,

```ruby
if (void_next = body.each_node(:next).find { |n| n.children.empty? })
  add_offense(...)
end
```

**Comment by mvidner** [comment]
Thanks for the suggestions!

**Comment by bbatsov** [comment]
Overall the cop looks good. Just squash those commits together and update the final message to match our commit message style. (and rebase of top of `master`)

File: spec/rubocop/cop/lint/next_without_accumulator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
+# encoding: utf-8
+
+require 'spec_helper'
+
+describe RuboCop::Cop::Lint::NextWithoutAccumulator do
+  subject(:cop) { described_class.new }
+
+  context 'given a reduce block' do
```
Comment (bbatsov) [inline spec/rubocop/cop/lint/next_without_accumulator_spec.rb]: You can use shared examples for `reduce` and `inject`.

File: spec/rubocop/cop/lint/next_without_accumulator_spec.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    it 'registers an offense for a bare next' do
+      inspect_source(cop, '(1..4).reduce(0) {|a, i| next if i.odd?; a + i }')
+      expect(cop.highlights).to eq(['next'])
```
Comment (bbatsov) [inline spec/rubocop/cop/lint/next_without_accumulator_spec.rb]: I'd still add here a check there's one offence detected just to be on the safe side.

**Comment by bbatsov** [comment]
Drop the `.` from the commit message's title, please. Titles are not complete sentences.

**Comment by bbatsov** [comment]
Now the build seems to be failing.

**Comment by mvidner** [comment]
I've fixed the build. Thanks for the comments!

**Pull request merged**